linux/hci/hci.go

```diff
@@ -1001,3 +1018,44 @@ func (h *HCI) findConnection(handle uint16) *Conn {
 
 	return c
 }
+
+type opCodeLocker struct {
+	locks map[int]*sync.Mutex
+	sync.RWMutex
+}
+
+func (o *opCodeLocker) LockOpCode(oc int) {
+	o.Lock()
+
+	if o.locks == nil {
+		o.locks = map[int]*sync.Mutex{oc: {}}
+	}
+
+	var l *sync.Mutex
+	if lock, ok := o.locks[oc]; !ok {
+		o.locks[oc] = &sync.Mutex{}
+		l = o.locks[oc]
+	} else {
+		l = lock
+	}
+
+	o.Unlock()
```

Comment:
I would move this up to the top as a defer. it also would probably safer that way since we would be locking the opcode's lock under the map lock.

Reply:
It would be better but then it causes an issue. If we hold that lock while trying to Lock the opcode mutex, then it causes a double lock situation.